sphinx_click/ext.py

```diff
@@ -438,74 +519,134 @@ def _generate_nodes(
+            hide_current_header = True
+
         if semantic_group:
-            lines = _format_description(ctx)
+            lines = itertools.chain(lines, _format_description(ctx))
         else:
-            lines = _format_command(ctx, nested, commands)
+            lines = itertools.chain(
+                lines, _format_command(ctx, nested, commands, hide_current_header)
+            )
 
         for line in lines:
             LOG.debug(line)
             result.append(line, source_name)
 
-        sphinx_nodes.nested_parse_with_titles(self.state, result, section)
-
         # Subcommands
 
+        subcommand_nodes = []
         if nested == NESTED_FULL:
             if isinstance(command, click.CommandCollection):
                 for source in command.sources:
-                    section.extend(
+                    subcommand_nodes.extend(
                         self._generate_nodes(
                             source.name,
                             source,
                             parent=ctx,
                             nested=nested,
                             semantic_group=True,
+                            hide_header=False,  # Hiding the header should not propagate to children
                         )
                     )
             else:
-                commands = _filter_commands(ctx, commands)
-                for command in commands:
+                # We use the term "subcommand" here but these can be main commands as well
+                for subcommand in _filter_commands(ctx, commands):
                     parent = ctx if not semantic_group else ctx.parent
-                    section.extend(
+                    subcommand_nodes.extend(
                         self._generate_nodes(
-                            command.name, command, parent=parent, nested=nested
+                            subcommand.name,
+                            subcommand,
+                            parent=parent,
+                            nested=nested,
+                            hide_header=False,  # Hiding the header should not propagate to children
                         )
                     )
 
-        return [section]
+        final_nodes = []
+        if hide_header:
+            final_nodes = subcommand_nodes
+
+            if nested == NESTED_NONE or nested == NESTED_SHORT:
+                section = nodes.paragraph()
+                self.state.nested_parse(result, 0, section)
+                final_nodes.insert(0, section)
+
+        else:
+            # Title
+
+            section = nodes.section(
+                '',
+                nodes.title(text=name),
+                ids=[nodes.make_id(ctx.command_path)],
+                names=[nodes.fully_normalize_name(ctx.command_path)],
+            )
+
+            sphinx_nodes.nested_parse_with_titles(self.state, result, section)
+
+            for node in subcommand_nodes:
+                section.append(node)
+            final_nodes = [section]
```

Comment:
I feel the behavior here is inconsistent depending on `:nested:` `short` / `full`.
With `short`, the headers are hidden as expected (left is without `hide-header`)

![Screenshot_20220805_152116](https://user-images.githubusercontent.com/43481553/183086728-73f442d4-3ad5-45c1-b744-26adcf678acd.png)

With `full`, the headers are still hidden, but the main command also disappears:
in the following example, the missing command is
> renku project [OPTIONS] COMMAND [ARGS]...

![Screenshot_20220805_152211](https://user-images.githubusercontent.com/43481553/183087148-68270165-4140-412c-ba62-9d93b97ea818.png)

I thought this could be circumvented by combining  `:nested: none` and `:hide-header:`, but the result is slightly different since part of the header is still shown:
in the following example, the additional text is
> Project commands.

![Screenshot_20220805_152657](https://user-images.githubusercontent.com/43481553/183087539-af27f087-b6f8-4728-b617-c2e07013b48d.png)